File: 2020/Day10.py
def part_1(input):
    diffs = [b-a for a,b in zip(input, input[1:])]
    return diffs.count(1) * diffs.count(3)

def part_2(input):
    ways = {0: 1}
    for j in input[1:]:
        ways[j] = ways.get(j-1,0) + ways.get(j-2,0) + ways.get(j-3,0)
    return ways[input[-1]]

with open("2020/Day10_input.txt") as f:
    input = sorted(int(x) for x in f.read().split())
    input = [0] + input + [input[-1]+3]

    print(f"Part 1: {part_1(input)}")
    print(f"Part 2: {part_2(input)}")

File: 2022/Day10.go
package main

import (
	"fmt"
	"os"
	"strconv"
	"strings"
)

func run(in string) []int {
	x, xs := 1, []int{}
	for _, l := range strings.Split(in, "\n") {
		xs = append(xs, x)
		if strings.HasPrefix(l, "addx") {
			n, _ := strconv.Atoi(l[5:])
			xs = append(xs, x)
			x += n
		}
	}
	return xs
}

func part1(in string) int {
	xs, total := run(in), 0
	for c := 20; c <= 240 && c <= len(xs); c += 40 {
		total += c * xs[c-1]
	}
	return total
}

func part2(in string) string {
	xs, out := run(in), strings.Builder{}
	for i, x := range xs {
		if i%40 == 0 {
			out.WriteByte('\n')
		}
		if d := i%40 - x; d >= -1 && d <= 1 {
			out.WriteByte('#')
		} else {
			out.WriteByte('.')
		}
	}
	return out.String()
}

func main() {
	b, _ := os.ReadFile("2022/Day10_input.txt")
	in := strings.TrimRight(string(b), "\n")
	fmt.Println("Part 1:", part1(in))
	fmt.Println("Part 2:", part2(in))
}
